[quant][pt2e] Add early prototype top level quantize_pt2e APIs

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90802

Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* torch.ao.quantization.quantize_pt2e.prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* torch.ao.quantization.quantize_pt2e.convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Also added a backend config for the qnnpack_pt2e backend:
* torch.ao.quantization.backend_config.get_qnnpack_pt2e_backend_config

Note: everything related to quantize_pt2e are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Add early prototype top level quantize_pt2e APIs

Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Note: everything related to quantize_pt2 are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Note: everything related to quantize_pt2 are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Note: everything related to quantize_pt2 are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

qihqi commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -0,0 +1,205 @@
+from collections import OrderedDict
+import operator
+
+# TODO: longer term, don't need to retrace or parse the string
```
chagne TODO to TODO(qihan) so I will find it.

jerryzh168 replied:
OK, I also plan to rebase after Sherlock land his PR: https://github.com/pytorch/pytorch/pull/90803, we may not need this

qihqi commented on torch/ao/quantization/backend_config/_pt2e.py:
```diff
@@ -0,0 +1,151 @@
+    return binary_op_configs
+
+def get_pt2e_backend_config():
+    return BackendConfig("pytorch_2.0_export")  \
```
nit: remove \ (just put the whole thing in another parethesis)

qihqi commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1204,11 +1206,15 @@ def insert_observers_for_model(
             this_node_dtype_info = node_name_to_target_dtype_info[node.name]
-            output_not_a_tensor = this_node_dtype_info is None
+            if hasattr(node, "meta") and "val" in node.meta:
+                output_is_a_tensor = this_node_dtype_info is not None and \
```
nit: remove \

qihqi reviewed (approved):
change unit test to have asserts instead of prints

vkuzo commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -0,0 +1,205 @@
+
+# TODO: longer term, don't need to retrace or parse the string
+# we should have node.meta["nn_module_stack"] that store the dict
+def _infer_nn_stack_trace_and_append_on_mode(m, gm, args_as_list):
```
nit: can we move the utils to a utils file, and keep this file to user facing functions only?

jerryzh168 replied:
sure

vkuzo commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -0,0 +1,205 @@
+    model.graph.eliminate_dead_code()
+    model.graph.lint()
+
+def prepare_pt2e(
```
can this function do a check that `model` has to be an output of torchdynamo export, and it's not expected to work on any GraphModule?

jerryzh168 replied:
sounds good, not sure how we can do this, @qihqi @SherlockNoMad is there a way to check this now?

vkuzo replied:
probably not essential now, ok to punt IMO

vkuzo commented on torch/ao/quantization/backend_config/_pt2e.py:
```diff
@@ -0,0 +1,151 @@
+    return binary_op_configs
+
+def get_pt2e_backend_config():
+    return BackendConfig("pytorch_2.0_export")  \
```
PT2.0 export is not a backend in the way that backend_config is defined, should this be something like "qnnpack_pt2e" instead?

jerryzh168 replied:
hmmm, it's not pure qnnpack though, we may also have some quantized ops (e.g. quantized add, embedding_byte) that is run with executorch runtime

vkuzo replied:
it might not be pure qnnpack, but it's also definitely not a backend :)

If I understand the `backend_config` design correctly, each entry in a config should be for some hardware target + kernel library, for example "executorch runtime with QNNPACK".  Neither PT2.0 or executorch are tied to any single hardware target.

Can we make the name of the file and config more specific?

kimishpatel replied:
This is exactly why we need some separation between qconfig and backend config. I am assuming that it is easier to say that give me a qconfig for this backend, but can we also have api that has some predefined qconfig names. Such as symmetric, asymmetric, etc. Would that be any better?

You would do backend related compatibility checks elsewhere. To be sure, I understand the appeal of backend config and I dont have any better ideas to make this cleaner, but I do see the confusion arising when we mix runtime with backend and delegate.

jerryzh168 replied:
@vkuzo I'm not sure what would be a good name for this, I'll rename it to _qnnpack_pt2e.py for now

@kimishpatel how is this related to qconfig? also not sure what is the confusion since we have explained what qconfig and backend_config means multiple times

kimishpatel replied:
I am talking about the conversation we. had around relatiion of qconfig to backend config. Anyways, I dont want to drag this any further. Ok with whatever you guys decide.

vkuzo replied:
> @vkuzo I'm not sure what would be a good name for this, I'll rename it to _qnnpack_pt2e.py for now

I think "qnnpack_pt2e" is actually the correct name.  In the test case, we are comparing the results of this workflow with the FX graph mode quantization workflow with the qnnpack backend_config and expecting them to match.

jerryzh168 replied:
sounds good, my main point about this is that we also have some operators that is not run with "qnnpack" the library (e.g. embedding_byte), but "qnnpack" as a backend in pytorch ao means something different, it can contain these ops as well. it might be helpful to clarify these things as well I think

vkuzo replied:
that makes sense.  I think it's the same story with non-executorch usage of qnnpack, right?

I guess I'd prefer:
1. names are consistent with each other (if we have backend_config_qnnpack, the pt2e version should be backend_config_qnnpack_pt2e)
2. names make as much sense as possible (maybe we should rename backend_config_qnnpack to something more specific)

I was just commenting on (1), but (2) can also be a future improvement to all backend configs

jerryzh168 replied:
yeah agree, I think we just need to document what these means exactly

kimishpatel commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -0,0 +1,205 @@
+    # initialize with top level parent scope
+    nn_module_stack_renamed = OrderedDict([("", None)])
+    if nn_module_stack:
+        # Rename module_key, e.g. "self_layer1_1__conv1" to "self.layer1.1._conv1", for easier downstream parsing
```
How would this distinguish `self.module_1 = ...`. You would replace this with self.module.1, no? But that woudl be wrong. Basically _ can be part of the attribute name.

jerryzh168 replied:
this is fine, we will look at what is the previous key and use that to figure out the current module path, in the example you gave, it will be something like the following:
{"self_module_1": ..., "self_module_1_abc": ...}
when we parse the first key, we'll get "self.module_1", when we parse the second key, we'll have "self.module_1.abc"

kimishpatel commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -0,0 +1,205 @@
+                    new_key = key[5:]
+                    prev_key = new_key
+            else:
+                new_key = prev_key + "." + key[len(prev_key) + 6 :]
```
what is + 6 here for?

jerryzh168 replied:
I think this is "self_" + parent_module_fqn (prev_key) + + current_module_fqn + "_"
6 = len("self_") + len("_")

jerryzh168 replied:
ideally we can do this in the upstream, I hope @SherlockNoMad can help integrate this to dynamo directly

kimishpatel reviewed (commented):
Left some comments

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Note: everything related to quantize_pt2 are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -0,0 +1,76 @@
+import copy
+from typing import Tuple, Any, Dict
+
+# TODO: longer term, don't need to retrace or parse the string
```
this is not moved to utils.py yet since I hope to remove it after Sherlock's PR is landed

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* torch.ao.quantization.quantize_pt2e.prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* torch.ao.quantization.quantize_pt2e.convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Also added a backend config for the qnnpack_pt2e backend:
* torch.ao.quantization.backend_config.get_qnnpack_pt2e_backend_config

Note: everything related to quantize_pt2 are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* torch.ao.quantization.quantize_pt2e.prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* torch.ao.quantization.quantize_pt2e.convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Also added a backend config for the qnnpack_pt2e backend:
* torch.ao.quantization.backend_config.get_qnnpack_pt2e_backend_config

Note: everything related to quantize_pt2 are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* torch.ao.quantization.quantize_pt2e.prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* torch.ao.quantization.quantize_pt2e.convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Also added a backend config for the qnnpack_pt2e backend:
* torch.ao.quantization.backend_config.get_qnnpack_pt2e_backend_config

Note: everything related to quantize_pt2e are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* torch.ao.quantization.quantize_pt2e.prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* torch.ao.quantization.quantize_pt2e.convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Also added a backend config for the qnnpack_pt2e backend:
* torch.ao.quantization.backend_config.get_qnnpack_pt2e_backend_config

Note: everything related to quantize_pt2e are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add early prototype top level quantize_pt2e APIs"


Summary:
This PR introduces the top level APIs for quantization support in PyTorch 2.0 Export stack
* torch.ao.quantization.quantize_pt2e.prepare_pt2e
Takes a model that is captured by the PyTorch 2.0 export (torchdynamo full graph mode) and prepares the model for calibration
for post training quantization

* torch.ao.quantization.quantize_pt2e.convert_pt2e
Takes a calibrated model and converts that to a reference quantized model that can be lowered later to quantized operator libraries or delegation modules

Also added a backend config for the qnnpack_pt2e backend:
* torch.ao.quantization.backend_config.get_qnnpack_pt2e_backend_config

Note: everything related to quantize_pt2e are experimental (prototype), and we don't have any bc guarantees

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "somehow broke test_resnet18 (quantization.fx.test_quantize_pt2e.TestQuantizePT2EModels), see https://hud.pytorch.org/pytorch/pytorch/commit/a66af1feba90cc64381bec45b0aa20ec778c92c5" -c landrace
